### PR TITLE
Differentiation of REJECTED and FAILED deposit state

### PR DESCRIPTION
**Pull request by vesaakerman**
fixes EASY-1146

#### When applied it will
* throw `RejectedDepositException` when deposit state should be set to **REJECTED**
* gives in `RejectedDepositException` the message that will be relayed to the client

#### related pull requests on github
repo                       | PR
-------------------------- | -----------------
easy-ingest-dispatcher      | [41] 
easy-stage-dataset            | [97]

@DANS-KNAW/easy

### CODE REVIEW

File: src/main/scala/nl/knaw/dans/easy/ingest_flow/Execution.scala [hunk 1]
```diff
@@ -134,7 +139,7 @@ trait Execution extends PidRequester {
   }
 
   def setDepositStateToArchived(datasetPid: String): Try[Unit] = Try {
-    setDepositState(stateArchived, settings.datasetAccessBaseUrl + "/" + datasetPid)
+    setDepositState(ARCHIVED.toString, settings.datasetAccessBaseUrl + "/" + datasetPid)
```
Comment (janvanmansum) [inline src/main/scala/nl/knaw/dans/easy/ingest_flow/Execution.scala]: Doesn't it make more sense to do the `toString` *in* `setDepositState`?

File: src/main/scala/nl/knaw/dans/easy/ingest_flow/package.scala [hunk 2]
```diff
@@ -72,6 +68,7 @@ package object ingest_flow {
   case class JSONParseException(msg: String, cause: Throwable) extends Exception(msg, cause)
   case class FailingResponseException(msg: String) extends Exception(msg)
   case class InvalidUserException(userId: String) extends Exception(s"not allowed to ingest this dataset on behalf of depositor $userId")
+  case class RejectedDepositException(msg: String = "", cause: Throwable = null) extends Exception(msg, cause)
```
Comment (janvanmansum) [inline src/main/scala/nl/knaw/dans/easy/ingest_flow/package.scala]: Same comments as [here](https://github.com/DANS-KNAW/easy-stage-dataset/pull/97#pullrequestreview-8846268)

File: src/main/scala/nl/knaw/dans/easy/ingest_flow/Execution.scala [hunk 3]
```diff
@@ -73,15 +78,15 @@ trait Execution extends PidRequester {
...
+      .getOrElse(Failure(error("SDO-set didn't contain a dataset object.")))
   }
 
   def waitForFedoraSync(datasetPid: String, pidDictionary: PidDictionary): Try[Unit] = {
     val expectedPids = pidDictionary.values.toSet - datasetPid
     @tailrec def loop(n: Int): Try[Unit] = {
       log.info(s"Check whether Fedora is synced. Tries left: $n.")
       if (n <= 0)
-        Failure(new RuntimeException(s"Fedora didn't sync in time. Dataset: $datasetPid. Expected pids: $expectedPids"))
+        Failure(error(s"Fedora didn't sync in time. Dataset: $datasetPid. Expected pids: $expectedPids"))
```
Comment (janvanmansum) [inline src/main/scala/nl/knaw/dans/easy/ingest_flow/Execution.scala]: This is not going to work. `error` doesn't just *create* the `RuntimeException`, it also throws it.

File: src/main/scala/nl/knaw/dans/easy/ingest_flow/EasyIngestFlow.scala [hunk 4]
```diff
@@ -84,7 +89,9 @@ object EasyIngestFlow {
     log.info(s"Scanning for viruses: $cmd")
     val output = new StringBuilder
     val exit = Process(cmd) ! ProcessLogger(line => output ++= s"$line\n")
-    if (exit > 0) throw new RuntimeException(s"Detected a virus (or scanner terminated with an error), virus scanner output:\n${output.toString}")
+    if (exit > 0) {
+      throw RejectedDepositException("Detected a virus (or scanner terminated with an error)", new Throwable(s"Detected a virus (or scanner terminated with an error), virus scanner output:\n${output.toString}"))
```
Comment (vesaakerman) [inline src/main/scala/nl/knaw/dans/easy/ingest_flow/EasyIngestFlow.scala]: In the `RejectedDepositException` we can give the message and the throwable.  In most cases the message is enough, this is actually the only place where throwable is given. But it makes it possible to write in the log a more detailed cause, like in this case the output of the virus scanner

File: src/main/scala/nl/knaw/dans/easy/ingest_flow/Execution.scala (hunk 1 above)
Comment (vesaakerman) [inline src/main/scala/nl/knaw/dans/easy/ingest_flow/Execution.scala]: That is a good idea.

File: src/main/scala/nl/knaw/dans/easy/ingest_flow/package.scala (hunk 2 above)
Comment (vesaakerman) [inline src/main/scala/nl/knaw/dans/easy/ingest_flow/package.scala]: I agree

File: src/main/scala/nl/knaw/dans/easy/ingest_flow/Execution.scala (hunk 3 above)
Comment (vesaakerman) [inline src/main/scala/nl/knaw/dans/easy/ingest_flow/Execution.scala]: oeps... you are right

File: src/main/scala/nl/knaw/dans/easy/ingest_flow/EasyIngestFlow.scala (hunk 4 above)
Comment (rvanheest) [inline src/main/scala/nl/knaw/dans/easy/ingest_flow/EasyIngestFlow.scala]: I'd say we merge the two messages into one message in the `RejectedDepositException`. In my opinion it doesn't add much to have this extra `Throwable`.

File: src/main/scala/nl/knaw/dans/easy/ingest_flow/package.scala [hunk 5]
```diff
@@ -120,9 +118,9 @@ package object ingest_flow {
     settings.depositorId == "mendeleydata" || settings.depositorId == "mendeltest"
   }
 
-  def setDepositState(state: String, description: String)(implicit s: Settings): Try[Unit] = Try {
+  def setDepositState(state: State, description: String)(implicit s: Settings): Try[Unit] = Try {
     val stateFile = new PropertiesConfiguration(depositPropertiesFile)
-    stateFile.setProperty("state.label", state)
+    stateFile.setProperty("state.label", state.toString)
```
Comment (rvanheest) [inline src/main/scala/nl/knaw/dans/easy/ingest_flow/package.scala]: If you like, you don't even need the `toString` here. This parameter has type `Object` anyway!

File: src/main/scala/nl/knaw/dans/easy/ingest_flow/MendeleyExecution.scala [hunk 6]
```diff
@@ -65,23 +64,23 @@ abstract class MendeleyExecution(implicit override val settings: Settings) exten
   }
 
   def loadDDM() = Try {
-    datasetMetadata.map(XML.loadFile)
+    datasetMetadata.filter(_.exists).map(XML.loadFile)
       .getOrElse(error(s"Could not find bag in deposit: ${settings.depositDir}"))
   }
 
   def assertNoAccessSet(xml: Elem): Try[Unit] = Try {
```
Comment (rvanheest) [inline src/main/scala/nl/knaw/dans/easy/ingest_flow/MendeleyExecution.scala]: We can also do it with `Failure` and `Success`:

```scala
def assertNoAccessSet(xml: Elem): Try[Unit] = {
    (xml \\ "DDM" \ "profile" \ "accessRights").map(_.text) match {
      case Seq() => Failure(RejectedDepositException("Dataset metadata contains no access rights"))
      case Seq(ar) if ar == "NO_ACCESS" => Success(())
      case Seq(ar) => Failure(RejectedDepositException("Dataset DOI is other access but accessrights are NOT set to NO_ACCESS"))
      case multiple => Failure(RejectedDepositException(s"Dataset metadata contains multiple access rights: $multiple"))
    }
  }
```

File: src/main/scala/nl/knaw/dans/easy/ingest_flow/MendeleyExecution.scala [hunk 7]
```diff
@@ -65,23 +64,23 @@ abstract class MendeleyExecution(implicit override val settings: Settings) exten
...
-      case Seq() => error("Dataset metadata contains no access rights")
-      case Seq(ar) => if (ar != "NO_ACCESS") error("Dataset DOI is other access but accessrights are NOT set to NO_ACCESS")
-      case multiple => error(s"Dataset metadata contains multiple access rights: $multiple")
+      case Seq() => throw RejectedDepositException("Dataset metadata contains no access rights")
+      case Seq(ar) => if (ar != "NO_ACCESS") throw RejectedDepositException("Dataset DOI is other access but accessrights are NOT set to NO_ACCESS")
+      case multiple => throw RejectedDepositException(s"Dataset metadata contains multiple access rights: $multiple")
     }
   }
 
   def getDoiFromDdm(xml: Elem): Try[String] = Try {
```
Comment (rvanheest) [inline src/main/scala/nl/knaw/dans/easy/ingest_flow/MendeleyExecution.scala]: try the same as above yourself 😃

File: src/main/scala/nl/knaw/dans/easy/ingest_flow/State.scala [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package nl.knaw.dans.easy.ingest_flow
+
+/**
+ * Created by vesa on 11/11/16.
```
Comment (rvanheest) [inline src/main/scala/nl/knaw/dans/easy/ingest_flow/State.scala]: maybe remove this comment?

**Pull request merged**